implants/lib/transport/src/dns.rs

```diff
@@ -0,0 +1,1627 @@
+                let rdata_end = offset + rdlength as usize;
+                while offset < rdata_end && offset < response.len() {
+                    let txt_len = response[offset] as usize;
+                    offset += 1;
+                    if offset + txt_len <= response.len() && offset + txt_len <= rdata_end {
+                        record_data.extend_from_slice(&response[offset..offset + txt_len]);
+                        offset += txt_len;
+                    } else {
+                        break;
+                    }
+                }
+            } else if rtype == A_RECORD_TYPE || rtype == AAAA_RECORD_TYPE {
+                // A or AAAA record - extract IP address bytes
+                if offset + rdlength as usize <= response.len() {
+                    record_data.extend_from_slice(&response[offset..offset + rdlength as usize]);
+                    offset += rdlength as usize;
+                }
+            } else {
+                offset += rdlength as usize;
+            }
+        }
+
+        Ok(record_data)
+    }
+
+    /// Send a single DNS query and receive response, with record type fallback
+    async fn send_query(&mut self, subdomain: &str) -> Result<Vec<u8>> {
+        use rand::Rng;
+
+        let socket = self
+            .socket
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("Socket not initialized"))?;
+
+        // Determine which record types to try
+        let record_types_to_try: Vec<u16> = if self.enable_fallback {
+            // Try all record types in priority order, but start with preferred
+            let mut types = Vec::new();
+            types.push(self.preferred_record_type);
+            for &rt in RECORD_TYPE_PRIORITY {
+                if rt != self.preferred_record_type {
+                    types.push(rt);
+                }
+            }
+            types
+        } else {
+            // Only try the preferred record type
+            vec![self.preferred_record_type]
+        };
+
+        // Try each record type
+        for &record_type in &record_types_to_try {
+            #[cfg(debug_assertions)]
+            {
+                let type_name = match record_type {
+                    TXT_RECORD_TYPE => "TXT",
+                    A_RECORD_TYPE => "A",
+                    AAAA_RECORD_TYPE => "AAAA",
+                    _ => "UNKNOWN",
+                };
+                log::trace!("Attempting DNS query with record type: {}", type_name);
+            }
+
+            // Generate random transaction ID
+            let transaction_id: u16 = rand::thread_rng().gen();
+            let query = self.build_dns_query(subdomain, transaction_id, record_type);
+
+            // Determine DNS server to use
+            let target = if let Some(ref server) = self.dns_server {
+                server.clone()
+            } else {
+                // Use system resolver - send to localhost:53
+                "127.0.0.1:53".to_string()
```

Comment:
localhost:53 isn't always the system resolver.
This should work on newer linux but older linux and macos don't.
Check out hickory-reslover it may have an easy cross platform way to find the systems default resolver.